pkg/services/annotations/annotationsimpl/loki/alert_state_history_store.go

```diff
@@ -0,0 +1,399 @@
+package loki
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/services/annotations/accesscontrol"
+
+	"github.com/grafana/grafana/pkg/infra/db"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/grafana/grafana/pkg/services/ngalert/metrics"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+	"github.com/grafana/grafana/pkg/services/ngalert/state/historian"
+	historymodel "github.com/grafana/grafana/pkg/services/ngalert/state/historian/model"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
+
+// from grafana/pkg/services/ngalert/state/historian/loki.go
+const defaultQueryRange = 6 * time.Hour
+
+var (
+	ErrLokiStoreInternal    = errutil.NewBase(errutil.StatusInternal, "annotations.loki.internal")
+	ErrLokiStoreUnsupported = errutil.NotImplemented("annotations.loki.unsupported")
+	ErrMissingRule          = errors.New("missing rule")
+)
+
+type lokiQueryClient interface {
+	RangeQuery(ctx context.Context, query string, from, to, limit int64) (historian.QueryRes, error)
+}
+
+type AlertStateHistoryStore struct {
+	client lokiQueryClient
+	db     db.DB
+	log    log.Logger
+}
+
+func NewLokiAlertStateHistoryStore(cfg *setting.Cfg, db db.DB, log log.Logger) (*AlertStateHistoryStore, error) {
+	metrics := metrics.NewNGAlert(prometheus.DefaultRegisterer)
```

Comment:
Won't this create `duplicate metrics registration attempted` panics once this is actually constructed outside of test code? The same metrics are already registered in the ngalert service.

Reply:
Probably yes. I considered moving the single metric the client uses to its own struct or registering it within the client itself, WDYT about either of these?